render revealed fields as chips

saurfang opened the pull request:
#597 

Introduce a new RevealMapArgument sub type under ObjectArgument, which renders with Chip component. 

Before

<img width="482" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/4a0c3fb6-90c4-4f71-9bcf-4fe25ab92d0f">


After

<img width="482" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/e1dfe5a8-495f-4c67-a1c7-cfa7a85e61d6">


~p.s. I will make a separate PR to refactor argument description into a clickable tooltip where we can tell users more about how to interpret these inputs. We are inline description today but it is too verbose to be fit cleanly into the UI.~

I also plan to add a toggle of some sort so user can inspect the underlying raw input (e.g. the object JSON in this case.) But suffice to say, everyday user would immediately understand with this improved UI where they likely gloss over the raw json completely in the original UI.

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -379,6 +386,58 @@ export function ObjectArgInput({
   );
 }
 
+function RevealArgInput({
```
i think it would be good to rename this to something more general - eg 'togglelist' - something that lets you toggle a bunch of values independently. that way it could be reused by other PCDs as well, in case they need a toggle list that doesn't necessarily toggle things that mean 'reveal or not reveal'

ichub commented on packages/pcd-types/src/pcd.ts:
```diff
@@ -265,3 +265,20 @@ export type PCDArgument<T extends PCD = PCD> = Argument<
+  ArgumentTypeName.Object,
+  T
+>;
+export function isRevealMapArgument(
```
i think adding an explicit new type here, and via `pcd.ts` makes sense. That way it could also show up as a 'Toggle List' in the UI, rather than an 'Object'

ichub commented on apps/passport-client/components/core/Chip.tsx:
```diff
@@ -0,0 +1,59 @@
+ *
+ * See https://m3.material.io/components/chips/overview
+ */
+const Chip: React.FC<{
```
could you declare this functional component in a way consistent with the way the rest of them are defined in this project?

```
export function Chip({...}) ...
```

ichub reviewed (changes requested):
this is a great idea, and looks nice too! 

I have a few suggestions about how to make this fit into the framework a little better, so that it can be reused by other PCDs as well.

saurfang commented:
Renamed and added a new ArgumentType called ToggleList as suggested. There is the small tension on how to best represent the semantics of RevealFields. I am hard coding the frontend component for now and open to explore better options in the future. 

ptal @ichub 

<img width="375" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/1724f336-e176-4052-8aa3-6537e6c250e5">

ichub commented:
> There is the small tension on how to best represent the semantics of RevealFields

Yeah, I can see that now. I think rather than using the eye open/close, we can use a checkmark. Another thing is that the revealed fields should probably be green, rather than red.

saurfang commented:
My rationale was that users should pay more attention to inputs that are revealed, which making it green feels a little odd. I buy that "green" as in I give you permission to view it but worry it conveys "green as in good".

I made it more color neutral and use a "checked" prop instead of chip and IMO this works just as well for the revealed fields usecase. 

I also made an attempt to support revealed fields specifically because that is the only ToggleList argument we have and I want to optimize best experience on what we already have. Let me know what you think. I am open to drop the eye icons but feel strongly we should hide the "reveal" prefix which made it look very rough.

<img width="379" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/2cc25bc0-93b1-4197-8fb5-cbf31624ade6">

fwiw happy to defer to designer to iterate on the exact text/bg/border color and weight. IMO the unchecked color can use a bit more work.

ichub reviewed (approved):
OK, lgtm. (We don't have a designer 😌)

Make sure to test that this works with your dark mode PR once you merge, and also that it doesn't break any of the existing examples in `consumer-client`

saurfang commented:
　double checked all examples in consumer-client. updated rendered screenshot in PR description
